cpp/ql/test/query-tests/Likely Bugs/Arithmetic/IntMultToLong/IntMultToLong.c

```diff
@@ -92,3 +92,13 @@ void use_printf(float f, double d)
 size_t three_chars(unsigned char a, unsigned char b, unsigned char c) {
     return a * b * c; // at most 16581375
 }
+
+void g(unsigned char a, unsigned char b, unsigned char b2, int c) {
+    unsigned long d, e, f, g, h;
+    d = (a + 1) * (b + 1); // GOOD
+    e = (c + 1) * (b + 1); // BAD
+    h = (a + 1) * (b + 1) * (b2 + 1); // GOOD
+
+    f = (a + (a + 1)) * (b + 1); // GOOD
+    g = (c + (a + 1)) * (b + 1); // BAD
+}
```

Comment:
I don't think these tests are pushing the query very hard. How about these:
```
g = (a + 1073741824) * b; // BAD [NOT DETECTED?]
g = (a + (1 << 30)) * b; // BAD
g = a * a * a * b * b * b; // BAD [NOT DETECTED?]
g = (a + (unsigned short)(-1)) * (b + (unsigned short)(-1)); // BAD [NOT DETECTED?]
```

Handling all these cases would require a different approach than what you have here. Computing the range of arbitrary subexpressions is hard, but our `SimpleRangeAnalysis` library does exactly that. The problem is that it doesn't handle multiplications because [it broke performance to support multiplication inside the main recursion of the range analysis](https://git.semmle.com/Semmle/code/pull/17898). But it should perform just fine to do a conservative analysis of multiplications just for this query _after_ the range analysis has finished.

Reply:
I'll include these tests (and more challenging ones going forward), and see if I can reduce the false negatives using the `SimpleRangeAnalysis`. The final test you included is currently detected, but the others aren't - and they definitely should be!